connectors/migrations/20241030_fix_notion_parents.ts

```diff
@@ -0,0 +1,127 @@
+  connector: ConnectorResource,
+  execute = false
+) {
+  let pagesIdCursor = 0;
+  let databasesIdCursor = 0;
+
+  const pageSize = 512;
+  let nodes: (NotionPage | NotionDatabase)[] = [];
+  do {
+    const pages = await NotionPage.findAll({
+      where: {
+        connectorId: connector.id,
+        parentId: "unknown",
+        id: {
+          [Op.gte]: pagesIdCursor,
+        },
+      },
+      limit: pageSize,
+      order: [["id", "ASC"]],
+    });
+    const databases = await NotionDatabase.findAll({
+      where: {
+        connectorId: connector.id,
+        parentId: "unknown",
+        id: {
+          [Op.gte]: databasesIdCursor,
+        },
+      },
+      limit: pageSize,
+      order: [["id", "ASC"]],
+    });
+
+    if (pages.length > 0) {
+      const newCursor = pages[pages.length - 1]?.id;
+      if (!newCursor) {
+        throw new Error("Last page is undefined");
+      }
+      pagesIdCursor = newCursor;
+    }
+    if (databases.length > 0) {
+      const newCursor = databases[databases.length - 1]?.id;
+      if (!newCursor) {
+        throw new Error("Last database is undefined");
+      }
+      databasesIdCursor = newCursor;
+    }
+
+    nodes = [...pages, ...databases];
+    const res = await concurrentExecutor(
+      nodes,
+      async (node) => {
+        const parents = await getParents(
+          connector.id,
+          "notionPageId" in node ? node.notionPageId : node.notionDatabaseId,
+          [],
+          undefined
+        );
+
+        let documentId: string | null = null;
+        let tableId: string | null = null;
+
+        if ("notionPageId" in node) {
+          // its a page
+          documentId = `notion-${node.notionPageId}`;
+        } else {
+          tableId = `notion-${node.notionDatabaseId}`;
+          documentId = `notion-database-${node.notionDatabaseId}`;
+        }
+
+        if (documentId) {
+          if (execute) {
+            await updateDocumentParentsField({
+              dataSourceConfig: dataSourceConfigFromConnector(connector),
+              documentId,
+              parents,
+            });
+          }
+        }
+
+        if (tableId) {
+          if (execute) {
+            await updateTableParentsField({
+              dataSourceConfig: dataSourceConfigFromConnector(connector),
+              tableId,
+              parents,
+            });
+          }
+        }
+      },
+      { concurrency: 8 }
+    );
+
+    console.log(
+      `Processed ${res.length} nodes, (pages cursor: ${pagesIdCursor}, databases cursor: ${databasesIdCursor})`
+    );
+  } while (nodes.length > 0);
+
+  console.log(
+    `Finished processing connector ${connector.id} (workspace ${connector.workspaceId})`
+  );
+}
+
+makeScript({}, async ({ execute }) => {
+  const connectors = await ConnectorResource.listByType("notion", {});
+
+  console.log(`Found ${connectors.length} Notion connectors`);
+  for (const connector of connectors) {
```

Comment:
Can we use concurrent executor to speed things up?

Reply:
SO it would be 2 levels of concurrent executor ? I think we don't have that many notion connectors so only one level of concurrent execution seems fine